src/CharsetDetector.cs

```diff
@@ -133,6 +133,37 @@ public static DetectionResult DetectFromBytes(byte[] bytes)
             return detector.DataEnd();
         }
 
+        /// <summary>
+        /// Detect the character encoding form this byte array.
+        /// </summary>
+        /// <param name="bytes"></param>
+        /// <param name="offset"></param>
+        /// <param name="len"></param>
+        /// <returns></returns>
+        public static DetectionResult DetectFromBytes(byte[] bytes, int offset, int len)
+        {
+            if (bytes == null)
+            {
+                throw new ArgumentNullException(nameof(bytes));
+            }
+            if (offset < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(offset));
+            }
+            if (len < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(len));
+            }
+            if (bytes.Length - offset < len)
+            {
+                throw new ArgumentOutOfRangeException(nameof(len));

```

Comment:
This exception is similar to the previous, but they are different

Reply:
My intention was to find the best exception without introducing additional strings.

.Net Framework's [FileStream.Write](https://github.com/microsoft/referencesource/blob/17b97365645da62cf8a49444d979f94a59bbb155/mscorlib/system/io/filestream.cs#L1758) implementation use `ArgumentException`.

```csharp
if (array.Length - offset < count)
	throw new ArgumentException(Environment.GetResourceString("Argument_InvalidOffLen"));
```

Should we benchmark the code and use `throw new ArgumentException("Invalid offset and length")`?

Reply:
It seems to me that `ArgumentOutOfRangeException` looks here, but @304NotModified will tell you what message to write, since I only help to conduct a review.

For example, [here](https://github.com/microsoft/referencesource/blob/08b84d13e81cfdbd769a557b368539aac6a9cb30/mscorlib/mscorlib.txt#L491) is a message in mscorelib.

And then, ```offset + len > bytes.Length``` looks better. Because `offset` may be larger than `bytes.Length`